jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -379,7 +409,7 @@ private void buildImages() {
                         () ->
                             new BuildImageStep(
                                     buildContext,
-                                    childProgressDispatcherFactory,
+                                    progressDispatcher.newChildProducer(),
```

Comment:
https://github.com/GoogleContainerTools/jib/pull/2699#issuecomment-673148403

Almost all of classes are immutable, so we didn't have to think about synchronization when accessing instances in the multi-threaded environment. However, `progressDispatcher` is an exception (understandably as a class to report progress updates to the user).

So, it is unsafe to access `progressDispatcher` from multiple threads. That is, calling `newChildProducer()` here needs synchronization; it's being called in multiple threads launched through the loop.

The simplest solution I think is to call `newChildProducer()` outside `executorService.submit()` on line 408.

Reply:
I see so that means creating a variable to hold the 'progressDispatcher.newChildProducer()` outside of the loop and then use the variable instead in the threads?

Reply:
Yes. You can actually see it's what the current code does in `StepsRunner`. For example,

```java
  private void pushApplicationLayers() {
    ProgressEventDispatcher.Factory childProgressDispatcherFactory =
        Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();

    results.applicationLayerPushResults =
        executorService.submit(
            () ->
                scheduleCallables(..., childProgressDispatcherFactory, ...)
  }
```

You won't need `Verify.verifyNotNull()` in your case. For naming, since `childProgressDispatcherFactory` is already taken, I think `progressDispatcherFactory` works.

Reply:
Yes from the looks of it ,its creating a `newProducer() ` out side of the `executorSubmiter`